# Exploring Gun Deaths in the US

## Load Dataset 

In [1]:
# Read the dataset in as a list using the csv module.
# Import the csv module.
# Open the file using the open() function.
# Use the csv.reader() function to load the opened file.
# Call list() on the result to get a list of all the data in the file.
# Assign the result to the variable data.
# Display the first 5 rows of data to verify everything.

In [2]:
import csv
file = open("guns.csv","r")
csv_reader = csv.reader(file)
data = list(csv_reader)
data[0:5]

[['',
  'year',
  'month',
  'intent',
  'police',
  'sex',
  'age',
  'race',
  'hispanic',
  'place',
  'education'],
 ['1',
  '2012',
  '01',
  'Suicide',
  '0',
  'M',
  '34',
  'Asian/Pacific Islander',
  '100',
  'Home',
  '4'],
 ['2', '2012', '01', 'Suicide', '0', 'F', '21', 'White', '100', 'Street', '3'],
 ['3',
  '2012',
  '01',
  'Suicide',
  '0',
  'M',
  '60',
  'White',
  '100',
  'Other specified',
  '4'],
 ['4', '2012', '02', 'Suicide', '0', 'M', '64', 'White', '100', 'Home', '4']]

## Prepare Dataset for Analysis

In [3]:
# Extract the first row of data, and assign it to the variable headers.
# Remove the first row from data.
# Display headers.
# Display the first 5 rows of data to verify that you removed the header row properly.

In [4]:
headers = data[0]
headers

['',
 'year',
 'month',
 'intent',
 'police',
 'sex',
 'age',
 'race',
 'hispanic',
 'place',
 'education']

In [5]:
data = data[1:]
data[0:5]

[['1',
  '2012',
  '01',
  'Suicide',
  '0',
  'M',
  '34',
  'Asian/Pacific Islander',
  '100',
  'Home',
  '4'],
 ['2', '2012', '01', 'Suicide', '0', 'F', '21', 'White', '100', 'Street', '3'],
 ['3',
  '2012',
  '01',
  'Suicide',
  '0',
  'M',
  '60',
  'White',
  '100',
  'Other specified',
  '4'],
 ['4', '2012', '02', 'Suicide', '0', 'M', '64', 'White', '100', 'Home', '4'],
 ['5',
  '2012',
  '02',
  'Suicide',
  '0',
  'M',
  '31',
  'White',
  '100',
  'Other specified',
  '2']]

## Count Gun Deaths by Year

In [6]:
# Use a list comprehension to extract the year column from data.
# Because the year column is the second column in the data, you'll need to get the element at index 1 in each row.
# Assign the result to the variable years.
# Create an empty dictionary called year_counts.
# Loop through each element in years.
# If the element isn't a key in year_counts, create it, and set the value to 1.
# If the element is a key in year_counts, increment the value by one.
# Display year_counts to see how many gun deaths occur in each year.

In [7]:
years=[row[1] for row in data]
unique_years=set(years)
print(unique_years)
print(len(years))

{'2014', '2012', '2013'}
100798


In [8]:
year_counts=dict()
for row in years:
    if row in year_counts:
        year_counts[row]=year_counts[row]+1
    else:
        year_counts[row]=1
print(year_counts)

{'2014': 33599, '2012': 33563, '2013': 33636}


## Count Gun Deaths by Month and Year

In [9]:
# Use a list comprehension to create a datetime.datetime object for each row. Assign the result to dates.
# The year column is in the second element in each row.
# The month column is the third element in each row.
# Make sure to convert year and month to integers using int().
# Pass year, month, and day=1 into the datetime.datetime() function.
# Display the first 5 rows in dates to verify everything worked.
# Count up how many times each unique date occurs in dates. Assign the result to date_counts.
# This follows a similar procedure to what we did in the last screen with year_counts.
# Display date_counts.

In [10]:
import datetime
dates=[datetime.datetime(year=int(row[1]),month=int(row[2]),day=1)for row in data]
dates[0:5]

[datetime.datetime(2012, 1, 1, 0, 0),
 datetime.datetime(2012, 1, 1, 0, 0),
 datetime.datetime(2012, 1, 1, 0, 0),
 datetime.datetime(2012, 2, 1, 0, 0),
 datetime.datetime(2012, 2, 1, 0, 0)]

In [11]:
date_counts=dict()
for date in dates:
    if date in date_counts:
        date_counts[date]=date_counts[date]+1
    else:
        date_counts[date]=1
date_counts

{datetime.datetime(2012, 1, 1, 0, 0): 2758,
 datetime.datetime(2012, 2, 1, 0, 0): 2357,
 datetime.datetime(2012, 3, 1, 0, 0): 2743,
 datetime.datetime(2012, 4, 1, 0, 0): 2795,
 datetime.datetime(2012, 5, 1, 0, 0): 2999,
 datetime.datetime(2012, 6, 1, 0, 0): 2826,
 datetime.datetime(2012, 7, 1, 0, 0): 3026,
 datetime.datetime(2012, 8, 1, 0, 0): 2954,
 datetime.datetime(2012, 9, 1, 0, 0): 2852,
 datetime.datetime(2012, 10, 1, 0, 0): 2733,
 datetime.datetime(2012, 11, 1, 0, 0): 2729,
 datetime.datetime(2012, 12, 1, 0, 0): 2791,
 datetime.datetime(2013, 1, 1, 0, 0): 2864,
 datetime.datetime(2013, 2, 1, 0, 0): 2375,
 datetime.datetime(2013, 3, 1, 0, 0): 2862,
 datetime.datetime(2013, 4, 1, 0, 0): 2798,
 datetime.datetime(2013, 5, 1, 0, 0): 2806,
 datetime.datetime(2013, 6, 1, 0, 0): 2920,
 datetime.datetime(2013, 7, 1, 0, 0): 3079,
 datetime.datetime(2013, 8, 1, 0, 0): 2859,
 datetime.datetime(2013, 9, 1, 0, 0): 2742,
 datetime.datetime(2013, 10, 1, 0, 0): 2808,
 datetime.datetime(2013, 11,

In [12]:
print(max(date_counts.values()))
print(min(date_counts.values()))

3079
2357


## Explore Gun Deaths by Race and Sex

In [13]:
# Count up how many times each item in the sex column occurs.
# Assign the result to sex_counts.
# Count up how many times each item in the race column occurs.


In [14]:
sexes = [row[5] for row in data]
sex_counts={}
for sex in sexes:
    if sex in sex_counts:
        sex_counts[sex]+=1
    else:
        sex_counts[sex]=1
sex_counts

{'F': 14449, 'M': 86349}

In [15]:
# Assign the result to race_counts.
# Display race_counts and sex_counts to verify your work, 
#and see if you can spot any patterns.

In [16]:
races = [row[7] for row in data]
race_counts={}
for race in races:
    if race in race_counts:
        race_counts[race]+=1
    else:
        race_counts[race]=1
race_counts

{'Asian/Pacific Islander': 1326,
 'Black': 23296,
 'Hispanic': 9022,
 'Native American/Native Alaskan': 917,
 'White': 66237}

Analysis so far: 
* Seems like males are more involved in gun deaths than females.
* Minorities such as black and hispanic seem to be the most highly affected
* The number of deaths throught different months of seems to be vary with values ranging between: 2357 -  3079

## Adding a Second Dataset to Analysis

In [17]:
# Read in census.csv, and convert to a list of lists. 
# Assign the result to the census variable.
# Display census to verify your work.

In [18]:
import csv
file2 = open("census.csv","r")
csv_reader2 = csv.reader(file2)
data2 = list(csv_reader2)
data2[0:5]

[['Id',
  'Year',
  'Id',
  'Sex',
  'Id',
  'Hispanic Origin',
  'Id',
  'Id2',
  'Geography',
  'Total',
  'Race Alone - White',
  'Race Alone - Hispanic',
  'Race Alone - Black or African American',
  'Race Alone - American Indian and Alaska Native',
  'Race Alone - Asian',
  'Race Alone - Native Hawaiian and Other Pacific Islander',
  'Two or More Races'],
 ['cen42010',
  'April 1, 2010 Census',
  'totsex',
  'Both Sexes',
  'tothisp',
  'Total',
  '0100000US',
  '',
  'United States',
  '308745538',
  '197318956',
  '44618105',
  '40250635',
  '3739506',
  '15159516',
  '674625',
  '6984195']]

## Compute Rates of Gun Deaths per Race

In [19]:
# Manually create a dictionary, mapping that maps each key from race_counts to the population count of the race from census.
# The keys in the dictionary should be Asian/Pacific Islander, Black, Native American/Native Alaskan, Hispanic, and White.
# In the case of Asian/Pacific Islander, you'll need to add the counts from census for Race Alone - Asian, and Race Alone - Native Hawaiian and Other Pacific Islander.
# Create an empty dictionary, race_per_hundredk.
# Loop through each key in race_counts.
# Divide the value associated with the key in race_counts by the value associated with the key in mapping.
# Multiply by 100000.
# Assign the result to the same key in race_per_hundredk.
# When you're done, race_per_hundredk should contain the rate of gun deaths per 100000 people for each racial category.
# Print race_per_hundredk to verify your work.

In [20]:
mapping = {'Asian/Pacific Islander': 1326 + 15159516 + 674625,
 'Black': 23296 + 40250635,
 'Hispanic': 9022 + 44618105,
 'Native American/Native Alaskan': 917 + 3739506,
 'White': 66237 + 197318956}

race_per_hundredk = dict()
for race,value in race_counts.items():
    race_per_hundredk[race] = (value / mapping[race]) * 100000

race_per_hundredk
    
    

{'Asian/Pacific Islander': 8.37360843226158,
 'Black': 57.843869276132985,
 'Hispanic': 20.216403354847376,
 'Native American/Native Alaskan': 24.51594378496764,
 'White': 33.55722837832116}

## Filter by Intent

In [21]:
# Extract the intent column using a list comprehension. The intent column is the fourth column in data.
# Assign the result to intents.
# Extract the race column using a list comprehension. The race column is the eighth column in data.
# Assign the result to races.
# Create an empty dictionary called homicide_race_counts
# Use the enumerate() function to loop through each item in races. The position should be assigned to the loop variable i, and the value to the loop variable race.
# Check the value at position i in intents.
# If the value at position i in intents is Homicide:
# If the key race doesn't exist in homicide_race_counts, create it.
# Add 1 to the value associated with race in homicide_race_counts.
# When you're done, homicide_race_counts should have one key for each of the racial categories in data. The associated value should be the number of gun deaths by homicide for that race.
# Perform the same procedure we did in the last screen using mapping on homicide_race_counts to get from raw numbers to rates per 100000.
# Display homicide_race_counts to verify your work.
# Write up your findings in a markdown cell.
# Write up any next steps you want to pursue with the data in a markdown cell.

In [24]:
intents=[row[3] for row in data]
# races= [row[7] for row in data]
# done in a previous step
homicide_race_counts=dict()
for i, race in enumerate(races):
    if intents[i] == "Homicide":
        homicide_race_counts[race]+=1
    if race not in homicide_race_counts:
        homicide_race_counts[race]=0
homicide_race_counts
    
race_per_hundredk_homicide = dict()
for race,value in homicide_race_counts.items():
    race_per_hundredk_homicide[race] = (value / mapping[race]) * 100000

race_per_hundredk_homicide

{'Asian/Pacific Islander': 3.53005061360047,
 'Black': 48.4432473204565,
 'Hispanic': 12.624608346398816,
 'Native American/Native Alaskan': 8.715591792692965,
 'White': 4.634086205240329}


## Analysis so far:
* Seems that gun related homicides in the US affect more people in the Black and Hispanic races.

Some areas to investigate further:
* The link between month and homicide rate.
* Homicide rate by gender.
* The rates of other intents by gender and race.
* Gun death rates by location and education.